overlord/ifacestate/hotplug.go

```diff
@@ -86,14 +106,221 @@ func defaultDeviceKey(devinfo *hotplug.HotplugDeviceInfo, keyVersion int) (strin
+	}
+
+	hotplugIfaces := m.repo.AllHotplugInterfaces()
+	gadgetSlotsByInterface := make(map[string][]*snap.SlotInfo)
+	if gadget != nil {
+		for _, gadgetSlot := range gadget.Slots {
+			if _, ok := hotplugIfaces[gadgetSlot.Interface]; ok {
+				gadgetSlotsByInterface[gadgetSlot.Interface] = append(gadgetSlotsByInterface[gadgetSlot.Interface], gadgetSlot)
+			}
+		}
+	}
+
+InterfacesLoop:
+	// iterate over all hotplug interfaces
+	for _, iface := range hotplugIfaces {
+		hotplugHandler := iface.(hotplug.Definer)
+
+		// ignore device that is already handled by a gadget slot
+		if gadgetSlots, ok := gadgetSlotsByInterface[iface.Name()]; ok {
+			for _, gslot := range gadgetSlots {
+				if pred, ok := iface.(hotplug.HandledByGadgetPredicate); ok {
+					if pred.HandledByGadget(devinfo, gslot) {
+						logger.Debugf("ignoring device with path %q, interface %q (handled by gadget slot %s)", devinfo.DevicePath(), iface.Name(), gslot.Name)
+						continue InterfacesLoop
+					}
+				}
+			}
+		}
+
+		proposedSlot, err := hotplugHandler.HotplugDeviceDetected(devinfo)
+		if err != nil {
+			logger.Noticef("cannot process hotplug event by the rule of interface %q: %s", iface.Name(), err)
+			continue
+		}
+		if proposedSlot == nil {
+			continue
+		}
+
+		// Check the key when we know the interface wants to create a hotplug slot, doing this earlier would generate too much log noise about irrelevant devices
+		key, err := deviceKey(devinfo, iface, defaultKey)
+		if err != nil {
+			logger.Noticef("cannot compute hotplug key for device with path %s: %s", devinfo.DevicePath(), err.Error())
+			continue
+		}
+		if key == "" {
+			logger.Noticef("no valid hotplug key provided by interface %q, device with path %s ignored", iface.Name(), devinfo.DevicePath())
+			continue
+		}
+
+		proposedSlot, err = proposedSlot.Clean()
+		if err != nil {
+			logger.Noticef("cannot validate hotplug slot proposed by interface %q: %v", iface.Name(), err.Error())
+			continue
+		}
+		if proposedSlot.Label == "" {
+			si := interfaces.StaticInfoOf(iface)
+			proposedSlot.Label = si.Summary
+		}
+
+		if !hotplugFeature {
+			logger.Noticef("hotplug device add event ignored, enable experimental.hotplug")
+			return
+		}
+
+		logger.Debugf("adding hotplug device with path %s for interface %q, hotplug key %s", devinfo.DevicePath(), iface.Name(), key)
+
+		seq, err := allocHotplugSeq(st)
+		if err != nil {
+			logger.Noticef("internal error: cannot handle hotplug device with path %s: %s", devinfo.DevicePath(), err)
+			continue
+		}
+
+		// enumeratedDeviceKeys is only available when enumerating devices
+		if m.enumeratedDeviceKeys != nil {
+			if m.enumeratedDeviceKeys[iface.Name()] == nil {
+				m.enumeratedDeviceKeys[iface.Name()] = make(map[string]bool)
+			}
+			m.enumeratedDeviceKeys[iface.Name()][key] = true
+		}
+		devPath := devinfo.DevicePath()
+		m.hotplugDevicePaths[devPath] = append(m.hotplugDevicePaths[devPath], deviceData{hotplugKey: key, ifaceName: iface.Name()})
```

Comment:
this is a slice because we can have the same device under different interfaces at the same path? do we need to check that the interfaces are different?

Reply:
Correct, it's to handle different interfaces at same paths.

I don't think we need to check against duplicated interfaces here - we shouldn't have two "adds" for same path and interface without a "remove" in between; on "remove" we drop hotplugDevicePaths[devPath] entry.